### PR TITLE
chore: Add UnboundColumn to carry datatype for unbound reference

### LINKED ISSUE #511: Incorrect return type on aggregate functions implemented by AggregateUDF when upgrading to latest DataFusion
### Describe the bug

While working on https://github.com/apache/datafusion-comet/pull/403, one new test error happens:

```
- first/last *** FAILED *** (291 milliseconds)
  org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 17002.0 failed 1 times, most recent failure: Lost task 0.0 in stage 17002.0 (TID 49989) (1d885a59c17c executor driver): org.apache.comet.CometNativeException: Arrow error: Invalid argument error: column types must match schema types, expected Null but found Int32 at column index 0
```

The root cause of it is new design of aggregate expressions in DataFusion. Some aggregate expressions are now implemented using `AggregateUDF`.  To create the physical aggregate expression (i.e., `AggregateFunctionExpr `) for them, we need to call `create_aggregate_expr` API. But there is a circular relationship on the required arguments: https://github.com/apache/datafusion/issues/10785. It makes Comet cannot determine correct `input_phy_exprs` so an incorrect return type is determined during the call to `create_aggregate_expr`.

To fix it, we may need to revamp the design in DataFusion.

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #511.

### CODE REVIEW

**Comment by viirya** [comment]
This is proposed to fix the issue in #403 when upgrading to latest DataFusion.

**Comment by viirya** [comment]
cc @sunchao @andygrove @huaxingao

**Review by huaxingao — APPROVED** [review]

File: core/src/execution/datafusion/expressions/unbound.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    /// Get the data type of this expression, given the schema of the input
+    fn data_type(&self, _input_schema: &Schema) -> Result<DataType> {
+        Ok(self.datatype.clone())
+    }
+
+    /// Decide whehter this expression is nullable, given the schema of the input
```
Comment (huaxingao) [inline core/src/execution/datafusion/expressions/unbound.rs:76]: nit: `whehter` typo

Comment (viirya) [reply core/src/execution/datafusion/expressions/unbound.rs:76]: Oops

Comment (viirya) [reply core/src/execution/datafusion/expressions/unbound.rs:76]: I copied it from DataFusion. 😂

Comment (viirya) [inline core/src/execution/datafusion/expressions/unbound.rs:76]:
```suggestion
    /// Decide whether this expression is nullable, given the schema of the input
```

**Comment by viirya** [comment]
> LGTM.
> 
> Do we need to contribute the UnboundColumn back to DataFusion?

DataFusion doesn't need it.

**Comment by viirya** [comment]
Merged. Thanks @huaxingao @advancedxy

**Pull request merged**